pkg/proxy/handlers/spacelister.go

```diff
@@ -160,17 +163,55 @@ func (s *SpaceLister) ListUserWorkspaces(ctx echo.Context) ([]toolchainv1alpha1.
 		// account exists but the compliant username is not set yet, meaning it has not been fully provisioned yet, so return an empty list
 		return []toolchainv1alpha1.Workspace{}, nil
 	}
-
 	murName := signup.CompliantUsername
+
+	if workspaceName != "" {
+		// if a workspace was provided
+		// recursively get all the spacebindings, starting from this workspace and going up to the parent workspaces till the "root" of the workspace tree.
+		return s.hierarchicalWorkspaceSearch(ctx, workspaceName, signup)
+	}
+
+	// get all spacebindings with given mur since no workspace was provided
 	spaceBindings, err := s.listSpaceBindingsForUser(murName)
 	if err != nil {
 		ctx.Logger().Error(errs.Wrap(err, "error listing space bindings"))
 		return nil, err
 	}
-
 	return s.workspacesFromSpaceBindings(signup.Name, spaceBindings), nil
 }
 
+// hierarchicalWorkspaceSearch lists all spacebindings starting from a space and going to it's parentSpaces
+// if any of those spaces contains a SpaceBinding for the given user, then a workspace object will be returned with the found role.
+func (s *SpaceLister) hierarchicalWorkspaceSearch(ctx echo.Context, workspaceName string, signup *signup.Signup) ([]toolchainv1alpha1.Workspace, error) {
+	listSpaceBindingsFunc := func(spaceName string) ([]toolchainv1alpha1.SpaceBinding, error) {
+		spaceSelector, err := labels.NewRequirement(toolchainv1alpha1.SpaceBindingSpaceLabelKey, selection.Equals, []string{spaceName})
+		if err != nil {
+			return nil, err
+		}
+		return s.GetInformerServiceFunc().ListSpaceBindings(*spaceSelector)
+	}
+	spaceBindingLister := spacebinding.NewLister(listSpaceBindingsFunc, s.GetInformerServiceFunc().GetSpace)
+	space, err := s.GetInformerServiceFunc().GetSpace(workspaceName)
+	if err != nil {
+		ctx.Logger().Error(err, "failed to get space")
+		return nil, err
+	}
+	allSpaceBindings, err := spaceBindingLister.ListForSpace(space, []toolchainv1alpha1.SpaceBinding{})
+	if err != nil {
+		ctx.Logger().Error(err, "failed to list space bindings")
+		return nil, err
+	}
+	// check if user has access to this workspace
+	murName := signup.CompliantUsername
+	userBinding := filterUserSpaceBinding(murName, allSpaceBindings)
+	if userBinding == nil {
+		//  let's only log the issue and consider this as not found
+		ctx.Logger().Error(fmt.Sprintf("unauthorized access - there is no SpaceBinding present for the user %s and the workspace %s", murName, workspaceName))
+		return nil, nil
+	}
```

Comment:
is this really needed? the validation is already one in https://github.com/codeready-toolchain/registration-service/blob/3d66cb33031e62543a58332f27708fbf1c5e9394/pkg/proxy/proxy.go#L511-L524

Reply:
I've changed things a bit in https://github.com/codeready-toolchain/registration-service/pull/371/commits/796bed8496eb74d47f5ec2a1896cf47a85c87c02. I'm now using `GetUserWorkspace` which does that validation , but I need that so the function can return only 1 workspace (instead of all the workspaces of the user) which makes it makes it easier to test since I can directly call `GetUserWorkspace` from the hierarchical unit test cases.